ykrt/src/compile/jitc_yk/codegen/x64/mod.rs

```diff
@@ -2916,25 +2928,45 @@ impl CompiledTrace for X64CompiledTrace {
                 libc::PROT_WRITE | libc::PROT_EXEC,
             )
         };
+        // Check that the target address is 8-byte aligned. This is required so we can patch
+        // this atomically.
+        let patch_addr = patch_addr as *mut u64;
+        debug_assert!(patch_addr.addr() % 8 == 0);
         if res == 0 {
             // Calculate the relative offset from the patch address to the address of the
             // side-trace. We need to add the size of the jump instruction to the patch address.
-            // FIXME: Could we end up with an offset bigger than 32 bits?
-            let offset = i32::try_from(staddr as i64 - (patch_addr as i64 + JMP_32_SIZE)).unwrap();
-
-            // Check that the target address is 8-byte aligned. This is required so we can patch
-            // this atomically.
-            let patch_addr = patch_addr as *mut u64;
-            debug_assert!(patch_addr.addr() % 8 == 0);
-
-            // Generate the JMP instruction.
-            let mut bytes: [u8; 8] = [NOP_OPCODE; 8]; // Fill with NOPs
-            bytes[0] = JMP_OPCODE;
-            bytes[1..5].copy_from_slice(&offset.to_ne_bytes());
-            let pval = u64::from_ne_bytes(bytes);
-            unsafe {
-                // Now write the new instruction in a single atomic operation.
-                *patch_addr = pval;
+            let offset = i32::try_from(staddr as i64 - (patch_addr as i64 + JMP_32_SIZE));
+            if let Ok(offset) = offset {
+                // Emit a jump to a 32-bit offset.
+                let mut bytes: [u8; 8] = [NOP_OPCODE; 8]; // Fill with NOPs
+                bytes[0] = JMP_OPCODE;
+                bytes[1..5].copy_from_slice(&offset.to_ne_bytes());
+                let pval = u64::from_ne_bytes(bytes);
+                unsafe {
+                    // Now write the new instruction in a single atomic operation.
+                    *patch_addr = pval;
+                }
+            } else {
+                // Emit code to jump to a 64-bit address stored on the stack.
+                let addr_in_bytes = (staddr as u64).to_ne_bytes();
+                let mut pcode = Vec::new();
+                // Emit `mov qword [rsp-8], ...`
+                pcode.extend(&MOV_RSP_8);
+                pcode.extend(&addr_in_bytes[0..4]);
+                // Emit `mov qword [rsp-4], ...`
+                pcode.extend(&MOV_RSP_4);
+                pcode.extend(&addr_in_bytes[4..8]);
+                // Emit `jmp [rsp-8]`
+                pcode.extend(&JMP_RSP_8.to_be_bytes());
+                // Write the jump code after the old jump that skips the NOPs.
+                let patch_addr_64 = unsafe { (patch_addr as *mut u8).offset(8) };
+                unsafe {
+                    std::ptr::copy_nonoverlapping(pcode.as_ptr(), patch_addr_64, pcode.len());
```

Comment:
This is unsafe: at the very least it needs to be a sequence of atomic writes rather than a non-atomic copy.

However, that's not enough: it is possible that you write 8 bytes atomically, and then another thread executes this code. So you need to make sure that, after each atomic write, the "partially written" code can't do bad things. That will require a *very* careful comment justifying this.

Reply:
See this comment: https://github.com/ykjit/yk/pull/1591#discussion_r1952285268